apis/apiextensions/v1beta1/composition_types.go

```diff
@@ -133,35 +211,158 @@ type ReadinessCheck struct {
 	MatchInteger int64 `json:"matchInteger,omitempty"`
 }
 
-// Patch is used to patch the field on the base resource at ToFieldPath
-// after piping the value that is at FromFieldPath of the target resource through
-// transformers.
+// patchSetMap stores a list of patches from a set, against the Set ID.
+type patchSetMap map[string]Patch
+
+// PatchType defines the usable inbuilt patch types
+type PatchType string
+
+// Predefined PatchTypes.
+const (
+	PatchTypeFromCompositeFieldPath PatchType = "FromCompositeFieldPath" // Default
+	PatchTypePatchSet               PatchType = "PatchSet"
+)
+
+// Patch instances are applied between composite and composed resources.
+// Their behaviour depends on the Type value selected.
 type Patch struct {
+	// Type sets the patching behaviour to be used. Each patch type may require
+	// its' own fields to be set on the Patch object.
+	// +optional
+	// +kubebuilder:validation:Enum=FromCompositeFieldPath;PatchSet
+	// +kubebuilder:default=FromCompositeFieldPath
+	Type PatchType `json:"type,omitempty"`
+
+	// PatchSet allows other patch types to be nested under itself.
+	// PatchSets are defined on a Composition and can be referenced here,
+	// using their ID.
+	// +optional
+	*PatchSet `json:",inline"`
+
+	// PatchFromCompositeFieldPath is the 'standard' patch type that maps a field
+	// path from the composite resource to the composed resource, with optional
+	// transformations.
+	// +optional
+	*PatchFromCompositeFieldPath `json:",inline"`
```

Comment:
I'm not certain that we'll be able to cleanly embed/inline these different patch types in future, for example if `toFieldPath` and `fromFieldPath` were used by both `type: FromCompositeFieldPath` and `type: FromComposedFieldPath` (per the sketch at https://github.com/crossplane/crossplane/issues/1974).

Reply:
Given that `FromCompositeFieldPath` is the only patch type right now (prior to this PR), ISTM that inlining the fields only makes sense from a backwards-compatibility perspective, which is why I did it - I assume the existing patch style in `v1beta1` needs to continue working as-is unless a new API version was used.

My assumption was that any new patch types that require the same field names would work in the same manner as `Transform` currently does.

But I now realise I have already violated that as the `PatchSet` type is also inlined 🤦 Does it make sense for `PatchFromCompositeFieldPath` to be the only inlined patch type, for backwards compatibility, and everything else (including `PatchSet`) becomes a map item of the same name? (e.g. `type: PatchSet` -> `patchSet: { patchSetName: ... }`)